[14.0][IMP] l10n_br_fiscal: add operation line/product tags match

DiegoParadeda opened the pull request:
## Objetivo

Este PR introduz o modelo `product.tag` para aprimorar a capacidade de combinação automática das linhas de operação fiscal por produto.

mileo committed:
[IMP] l10n_br_fiscal: adds icms st demo data

DiegoParadeda committed:
[IMP] l10n_br_fiscal: use operation line/product  tags

DiegoParadeda committed:
[IMP] l10n_br_fiscal: use operation line/product  tags

DiegoParadeda committed:
[IMP] l10n_br_fiscal: add product.tag view

DiegoParadeda committed:
[IMP] l10n_br_fiscal: add product.tag view

mileo commented:
@DiegoParadeda pode implementar alguns testes e uns dados de demo?

marcelsavegnago commented:
De fato sinto falta de algo mais especifico para definir como filtro de  operação ou linha de operação fiscal, mas será que precisamos de mais um modelo ?

mileo commented:
> De fato sinto falta de algo mais especifico para definir como filtro de operação ou linha de operação fiscal, mas será que precisamos de mais um modelo ?

Poderia sugerir alguma outra opção? Das que pensamos e não fica legal:

- Botar mais boleanos / selections
- Alterar a regra de fiscal definition / fiscal op line;
- Usar o fiscal.tax.group, ou outro modelo do fiscal, mas acho que não é pra isso;

Essas tags dão mais versatilidade para casos complexos, onde existem mais de uma CFOP para a mesma operação, por exemplo a exportação com as CFOPs: 
- 7102 - Venda de mercadoria adquirida ou recebida de terceiros;
- 7501 - Exportação de mercadorias recebidas com fim específico de exportação;

A ideia é que o usuário nunca tenha que selecionar a linha da operação manualmente.

DiegoParadeda commented:
> De fato sinto falta de algo mais especifico para definir como filtro de operação ou linha de operação fiscal, mas será que precisamos de mais um modelo ?

@marcelsavegnago tentei usar a ideia dos marcadores (many2many) por simplicidade e praticidade apenas. Me pareceu ser a forma mais tranquila de cobrir vários casos de uso. Só queria ressaltar que o many2many na minha visão é importante.

Agora sobre o modelo em si, acho que daria pra herdar algum product.tags e colocar um campo nele exemplo "is_fiscal_tag" e usar esse modelo já existente com um domínio. Se acharem que fica melhor desse jeito posso alterar.

WesleyOliveira98 reviewed (approved):
LGTM

rvalyi commented:
@renatonlima pode revisar?

mileo reviewed (approved):
Pra mim isso já esta legal, mas o ideal é colocar testes para garantir que continua funcionando no futuro.

Tiago370 committed:
[IMP] l10n_br_pos: fix POS config test.

Tiago370 committed:
[IMP] l10n_br_fiscal: add product.tag test

Tiago370 committed:
[IMP] l10n_br_fiscal: add product.tag test

mileo commented on l10n_br_fiscal/tests/test_fiscal_document_generic.py:
```diff
@@ -291,6 +292,93 @@ def test_nfe_other_state(self):
+
+            # ICMS
+            if line.product_id.icms_origin not in ICMS_ORIGIN_TAX_IMPORTED:
+                self.assertEqual(
```
Pode fazer um assert da taxa do ICMS ST?

DiegoParadeda commented on l10n_br_fiscal/tests/test_fiscal_document_generic.py:
```diff
@@ -291,6 +292,93 @@ def test_nfe_other_state(self):
             )
 
+    def test_nfe_other_state_st(self):
+        """Test NFe other state."""
```
Pode atualizar a descrição do método por favor? Da pra melhorar um pouco os comentários também

Tiago370 replied:
Resolvido

DiegoParadeda commented on l10n_br_fiscal/tests/test_fiscal_document_generic.py:
```diff
@@ -291,6 +292,93 @@ def test_nfe_other_state(self):
+                self.assertEqual(
+                    line.cfop_id.code,
+                    "6404",
+                    "Error to mapping CFOP 6102"
```
O texto do assert está desatualizado aqui

Tiago370 replied:
Resolvido

DiegoParadeda commented on l10n_br_fiscal/tests/test_fiscal_document_generic.py:
```diff
@@ -291,6 +292,93 @@ def test_nfe_other_state(self):
+                    "Error to mapping CFOP 6102"
+                    " for Revenda de Contribuinte p/ Fora do Estado.",
+                )
+            # else:
```
Vai ter algum outro caso aqui? Senão já da pra remover esse trecho comentado

Tiago370 replied:
Resolvido

Tiago370 commented:
depend de https://github.com/OCA/l10n-brazil/pull/3422

DiegoParadeda commented on l10n_br_fiscal/tests/test_fiscal_document_generic.py:
```diff
@@ -291,6 +297,111 @@ def test_nfe_other_state(self):
+
+            # ICMS
+            if line.product_id.icms_origin not in ICMS_ORIGIN_TAX_IMPORTED:
+                line.write(
```
@Tiago370 acho que esse write não pode estar aqui. A ideia é testar o imposto automático. Se você escrever o imposto no teste a validação perde o sentido

DiegoParadeda commented on l10n_br_fiscal/tests/test_fiscal_document_generic.py:
```diff
@@ -291,6 +297,111 @@ def test_nfe_other_state(self):
+
+            # ICMS
+            if line.product_id.icms_origin not in ICMS_ORIGIN_TAX_IMPORTED:
+                line.write(
```
@Tiago370 acho que esse write não pode estar aqui. A ideia é testar o imposto automático. Se você escrever o imposto no teste a validação perde o sentido

DiegoParadeda commented on l10n_br_fiscal/tests/test_fiscal_document_generic.py:
```diff
@@ -291,6 +297,111 @@ def test_nfe_other_state(self):
+            line.with_company(empresa_lucro_presumido.id)._onchange_fiscal_taxes()
+
+            self.assertTrue("Revenda" in line.fiscal_operation_line_id.name)
+            if "Revenda" in line.fiscal_operation_line_id.name:
```
O assert da cfop_id precisa acontecer independente se "Revenda" está no nome da operation line. Essa linha é desnecessária

DiegoParadeda commented on l10n_br_fiscal/tests/test_fiscal_document_generic.py:
```diff
@@ -291,6 +297,111 @@ def test_nfe_other_state(self):
+            )._onchange_fiscal_operation_line_id()
+            line.with_company(empresa_lucro_presumido.id)._onchange_fiscal_taxes()
+
+            self.assertTrue("Revenda" in line.fiscal_operation_line_id.name)
```
Essa linha também é desnecessária já que esse assert já acontece algumas linhas abaixo em: "line.cfop_id.code, 6404"

Tiago370 commented:
Corrigido

mileo commented:
@DiegoParadeda @Tiago370 esta pronto para revisão?

Tiago370 commented:
@mileo da minha parte sim.

DiegoParadeda commented on l10n_br_fiscal/tests/test_fiscal_document_generic.py:
```diff
@@ -291,6 +297,98 @@ def test_nfe_other_state(self):
+            )
+
+            # ICMS
+            if line.product_id.icms_origin not in ICMS_ORIGIN_TAX_IMPORTED:
```
`if line.product_id.icms_origin not in ICMS_ORIGIN_TAX_IMPORTED:`

Esse if é desnecessário

DiegoParadeda commented on l10n_br_fiscal/tests/test_fiscal_document_generic.py:
```diff
@@ -291,6 +297,98 @@ def test_nfe_other_state(self):
+            )
+
+            # ICMS
+            if line.product_id.icms_origin not in ICMS_ORIGIN_TAX_IMPORTED:
```
Esse if é desnecessário

DiegoParadeda commented on l10n_br_fiscal/tests/test_fiscal_document_generic.py:
```diff
@@ -291,6 +297,98 @@ def test_nfe_other_state(self):
+                    " for Venda de Contribuinte p/ Fora do Estado.",
+                )
+                self.assertEqual(
+                    line.icmsst_tax_id.name,
```
Utilizar `registro.name` não é uma prática muito legal @Tiago370. Nesse caso você consegue pegar o XML ID do registro e garantir que o teste não vai quebrar se alguém trocar o nome do imposto

Tiago370 commented:
@DiegoParadeda corrigido.

marcelsavegnago commented:
ping @renatonlima

mileo commented:
@rvalyi @marcelsavegnago @antoniospneto @renatonlima podem revisar por gentileza, este pr é importante.

rvalyi commented:
teria que ver isso depois dos refatores para compute na v16. Tb teria que ver se é melhor no l10n_br_fiscal ou algum modulo adicional...

mileo committed:
[REF] l10n_br_fiscal: align the fiscal product tag field names

The tag was named differently on each side of the match
(product.operation_line_tag_ids against operation.line.product_fiscal_tag_ids),
which made the resolution domain hard to read. Both are now
fiscal_product_tag_ids, mirroring the product_tag_ids field Odoo itself
adds to product.template from 16.0 on, so the forward port does not
collide with it.

Also on the tag model: name becomes required, since Postgres accepts any
number of NULLs in a unique index and nameless tags would pile up; adds
active, so a tag can be retired without deleting it; orders by name; and
fixes the singular description.

mileo committed:
[IMP] l10n_br_fiscal: cover the operation line product tag resolution

The tag match added by this branch was only exercised end to end, through
demo documents and their onchange chain. The resolver itself
(_line_domain and _select_best_line) had no test at all.

Adds unit tests for the resolution rules: a tagged product reaches the
CFOP 6401/6404 lines, an untagged product never falls into a tagged line
(tagging restricts, it never widens), a tag used by no line degrades to
the plain line instead of returning nothing, matching is by intersection
when the product carries several tags, the documented score preference
for the tagged line, an archived tag stops pulling its line, the empty
candidate case and the tag name uniqueness and presence constraints.

renatonlima reviewed (changes requested):
O `l10n_br_fiscal` foi desenhado para que a parametrização de operações (CFOP, CST/CSOSN, NCM, CEST, NBS, grupo de tributação, tipo fiscal de operação, origem/destino, etc.) espelhe diretamente os **modelos de dados normativos** definidos pela legislação tributária: são tabelas publicadas e mantidas por SEFAZ/Receita Federal/CONFAZ (Tabela NCM/TIPI, Tabela CEST do Convênio ICMS 142/2018, Tabela CFOP do Ajuste SINIEF 07/01, tabelas de CST/CSOSN do Manual de Orientação do Contribuinte), e que aparecem literalmente nos campos do XML da NF-e/NFC-e e nos registros do SPED Fiscal. É esse alinhamento 1:1 entre "campo de parametrização" e "código validado pelo Fisco" que torna o motor de operações fiscais auditável e replicável entre implementações: qualquer contador ou consultor fiscal que abra a configuração de uma operação sabe exatamente o que cada campo significa, porque o significado vem da lei, não da empresa que configurou o sistema.

Uma tag genérica de produto quebra essa premissa em vários pontos:

1. **Não tem representação no documento fiscal.** Uma `product.tag` não existe no schema da NF-e nem em nenhum registro do SPED — é uma construção interna do Odoo, sem lastro em nenhuma tabela oficial. Isso significa que o campo que decide qual operação/CFOP será usado não pode ser conferido contra nenhuma fonte externa (malha fiscal, cruzamento de dados do Fisco, auditoria). Todo o resto do motor de parametrização foi pensado justamente para permitir essa rastreabilidade.

2. **É redundante com o que já existe.** O caso de uso citado no PR (ICMS-ST roteando para CFOP 6401 vs. 6404) é exatamente o problema que o **CEST** foi criado para resolver: identificar produtos sujeitos à substituição tributária. Se o motor de match de `operation.line` hoje não suporta filtrar por CEST (ou por faixa de NCM), o ajuste correto é estender esse motor para usar os campos normativos que já existem no produto — não introduzir uma segunda classificação paralela e não-oficial para contornar essa limitação.

3. **Quebra a interoperabilidade entre implementações.** Uma tabela como CFOP ou CEST tem o mesmo significado em qualquer instalação Brasil afora. Uma tag é vocabulário livre: cada implementador vai nomear e aplicar as tags do seu jeito, sem nenhuma fonte de verdade compartilhada. Isso é exatamente o oposto do que um módulo `l10n_*` mantido pela comunidade deveria fazer, que é convergir configurações usando o mesmo código-fonte de significado (a legislação), não fragmentar em taxonomias privadas por empresa.

4. **Risco de divergência silenciosa entre classificação fiscal real e tag manual.** Um produto pode ter NCM/CEST corretos, mas a tag ficar desatualizada ou mal aplicada — sem que exista mecanismo de validação cruzada, porque não há tabela externa contra a qual checar a tag (diferente de NCM/CEST, cuja inconsistência pode ser detectada comparando com a tabela oficial). Isso é uma porta aberta para erro de tributação sem alarme.

5. **Contexto agravante: a Reforma Tributária (EC 132/2023) está indo na direção oposta.** As Notas/Informes Técnicos da NF-e para IBS/CBS (IT 2025.002 e sucessoras) introduzem o **cClassTrib** — mais um código de classificação tributária normatizado e publicado oficialmente — justamente para tornar a definição de tratamento tributário ainda mais granular e codificada, não menos. Ou seja, a tendência regulatória é de mais campos oficiais e mais precisão nas tabelas, o que só reforça que o caminho do módulo deve ser mapear essas classificações normativas com mais fidelidade, e não introduzir uma camada paralela de tags livres exatamente no momento em que a complexidade normativa está aumentando.

mileo committed:
[FIX] l10n_br_fiscal: drop a call to a product only onchange in the ST tests

The new tax substitution tests call _onchange_ncm_id on the document
line, but that onchange belongs to l10n_br_fiscal.product.mixin, which
only product.template and product.product inherit, so the call raises
AttributeError. It only filled fiscal_genre_id from the NCM, which the
CFOP and CST assertions do not depend on.
